### PR TITLE
IGNITE-15785 POJO validation.

### PR DESCRIPTION
**Pull request by Berkof**
[IGNITE-15785](https://issues.apache.org/jira/browse/IGNITE-15785)

### CODE REVIEW

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/marshaller/RecordMarshallerTest.java [hunk 1]
```diff
@@ -424,9 +424,9 @@ public void ensureAllTypesChecked() {
                 new Column("primitiveByteCol", INT8, false, () -> (byte) 0x42),
                 new Column("primitiveShortCol", INT16, false, () -> (short) 0x4242),
                 new Column("primitiveIntCol", INT32, false, () -> 0x42424242),
-                new Column("primitiveFloatCol", FLOAT, false),
-                new Column("primitiveDoubleCol", DOUBLE, false),
-                
+                new Column("primitiveFloatCol", FLOAT, false, () -> 100.100),
```
Comment (korlov42) [inline modules/schema/src/test/java/org/apache/ignite/internal/schema/marshaller/RecordMarshallerTest.java:427]: Why did you change that?

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/Marshaller.java [hunk 2]
```diff
@@ -215,6 +232,10 @@ public Object readObject(Row reader) throws MarshallerException {
         @Override
         public void writeObject(Object obj, RowAssembler writer)
                 throws MarshallerException {
+            if (readOnly) {
```
Comment (korlov42) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/Marshaller.java:235]: Why do we need such a limitation? To me the previous behaviour is much better, because you got here an exception describing a root cause of the issue (setting null to a non-null field). Instead we now got here an abstract "Can't assemple row by object".

BTW there is a typo in "assemble".

Comment (korlov42) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/Marshaller.java:235]: Furthermore the current behaviour will be changed in case the user provide a null value supplier explicitly.

**Review by korlov42 — CHANGES_REQUESTED** [review]

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java [hunk 3]
```diff
@@ -96,16 +99,16 @@ public Column(
...
-     * @param defValSup   Default value supplier.
+     * @param defValSup   Default value supplier or {@code null} - if there is no default value supplier specified.
      */
     public Column(
             int columnOrder,
             String name,
             NativeType type,
             boolean nullable,
-            @NotNull Supplier<Object> defValSup
+            Supplier<Object> defValSup
```
[the remark this answers is not in the document]
Comment (Berkof) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java:109]: Remove from second public constructor. They will replace null with static null supplier and behaviour will be equal to SQL, where "default null" (like "() -> null" in Java) do nothing.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java [hunk 4]
```diff
@@ -178,6 +181,15 @@ public Object defaultValue() {
         return defValSup.get();
     }
 
+    /**
+     * Get no default value flag: {@code true} if column hasn't default value, {@code false} - otherwise.
+     *
+     * @return {@code true} if column hasn't default value, {@code false} - otherwise.
```
[the remark this answers is not in the document]
Comment (Berkof) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java:187]: Change logic after discussion. Now: no default value means that defValueSupplier returns null. No mapper - was it specified explicitly or implicitly.

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/SchemaDescriptorTest.java [hunk 5]
```diff
@@ -63,15 +63,15 @@ public void columnIndexedAccess() {
     @Test
     public void columnOrder() {
         Column[] keyColumns = {
-                new Column(0, "columnA", NativeTypes.INT8, false, () -> null),
-                new Column(1, "columnB", NativeTypes.UUID, false, () -> null),
-                new Column(2, "columnC", NativeTypes.INT32, false, () -> null),
+                new Column(0, "columnA", NativeTypes.INT8, false, null),
```
[the remark this answers is not in the document]
Comment (Berkof) [reply modules/schema/src/test/java/org/apache/ignite/internal/schema/SchemaDescriptorTest.java:66]: To use static null value supplier.

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/marshaller/RecordMarshallerTest.java (hunk 1 above)
Comment (Berkof) [reply modules/schema/src/test/java/org/apache/ignite/internal/schema/marshaller/RecordMarshallerTest.java:448]: Unnecessary change, reverted.

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/serializer/AbstractSerializerTest.java [hunk 6]
```diff
@@ -88,14 +88,14 @@ public void columnOrderSerializeTest() {
         AbstractSchemaSerializer assembler = SchemaSerializerImpl.INSTANCE;
 
         Column[] keyCols = {
-            new Column(0, "A", NativeTypes.UUID, false, () -> null),
-            new Column(1, "B", NativeTypes.INT64, false, () -> null),
-            new Column(2, "C", NativeTypes.INT8, false, () -> null),
+            new Column(0, "A", NativeTypes.UUID, false, null),
```
[the remark this answers is not in the document]
Comment (Berkof) [reply modules/schema/src/test/java/org/apache/ignite/internal/schema/serializer/AbstractSerializerTest.java:91]: To use static null value supplier.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/Marshaller.java (hunk 2 above)
Comment (Berkof) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/Marshaller.java]: Read only logic removed.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java [hunk 7]
```diff
@@ -30,6 +30,9 @@
  * user-defined order ({@link #columnOrder}).
  */
 public class Column implements Serializable {
+    /** Default "default value supplier". */
+    private static final Supplier<Object> NULL_SUPPLIER = () -> null;
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java:34]: NULL_SUPPLIER is not serializable now.
Seems, we can remove Serializable interface from this class,  SchemaDescriptor, and others as IGNITE-14679 already merged.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (AMashenkov) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java:187]: So, this method looks useless. If column default value wasn't configured then
For 'nullable' column - the default is implicit `null`.
For 'non-null' column - explicit null lead to invalid configuration error and implicit `null` means "no default wasn't set".

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/RecordMarshallerImpl.java [hunk 8]
```diff
@@ -59,11 +59,11 @@ public RecordMarshallerImpl(SchemaDescriptor schema, Mapper<R> mapper) {
         
         recClass = mapper.targetType();
         
-        keyMarsh = Marshaller.createMarshaller(schema.keyColumns().columns(), mapper);
+        keyMarsh = Marshaller.createMarshaller(schema.keyColumns().columns(), mapper, true);
         
         recMarsh = Marshaller.createMarshaller(
                 ArrayUtils.concat(schema.keyColumns().columns(), schema.valueColumns().columns()),
-                mapper
+                mapper, false
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/RecordMarshallerImpl.java:66]: This allows missed key columns for the record.

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/marshaller/KvMarshallerTest.java [hunk 9]
```diff
@@ -161,7 +161,14 @@ public void pojoWithFieldsOfAllTypes(MarshallerFactory factory) throws Marshalle
...
+        Column[] cols = new Column[]{
+                new Column("primitiveIntCol", INT32, false),
+                new Column("primitiveLongCol", INT64, false),
+                new Column("primitiveFloatCol", FLOAT, false),
+                new Column("primitiveDoubleCol", DOUBLE, false),
+                new Column("stringCol", STRING, false),
+                new Column("uuidCol", UUID, false),
+        };
         
         SchemaDescriptor schema = new SchemaDescriptor(1, cols, cols);
```
Comment (AMashenkov) [inline modules/schema/src/test/java/org/apache/ignite/internal/schema/marshaller/KvMarshallerTest.java:173]: Actually, this case is invalid.

Internally, we always know if column belong to key or value. But key and value columns with the same name can't mapped to a Record class.
So, we must fix and forbid this use case.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java (hunk 7 above)
Comment (Berkof) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java:33]: Remove Serializable from SchemaDescriptor, Columns, Column, NativeType, ColumnMapper, ColumnMapping.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java [hunk 10]
```diff
@@ -96,7 +115,7 @@ public Column(
      * @param name        Column name.
      * @param type        An instance of column data type.
      * @param nullable    If {@code false}, null values will not be allowed for this column.
-     * @param defValSup   Default value supplier.
+     * @param defValSup   Default value supplier or {@code null} - if there is no default value supplier specified.
```
Comment (korlov42) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java:118]: null is not allowed

Comment (Berkof) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/Column.java:118]: Rolled back.

**Review by korlov42 — APPROVED** [review]

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/marshaller/KvMarshallerTest.java [hunk 11]
```diff
@@ -161,7 +161,14 @@ public void pojoWithFieldsOfAllTypes(MarshallerFactory factory) throws Marshalle
...
     public void narrowType(MarshallerFactory factory) throws MarshallerException {
-        Column[] cols = columnsAllTypes();
+        Column[] cols = new Column[]{
+                new Column("primitiveIntCol", INT32, false),
+                new Column("primitiveLongCol", INT64, false),
+                new Column("primitiveFloatCol", FLOAT, false),
+                new Column("primitiveDoubleCol", DOUBLE, false),
+                new Column("stringCol", STRING, false),
+                new Column("uuidCol", UUID, false),
+        };
```
Comment (AMashenkov) [inline modules/schema/src/test/java/org/apache/ignite/internal/schema/marshaller/KvMarshallerTest.java:171]: Initially, test check we could map narrow type where amount of fields > amount of columns.
Now, amount of fields equal amount of columns.

Comment (Berkof) [reply modules/schema/src/test/java/org/apache/ignite/internal/schema/marshaller/KvMarshallerTest.java:171]: All types used for value.
